extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/customized/QuarkusJtaPlatform.java

```diff
@@ -1,26 +1,72 @@
 package io.quarkus.hibernate.orm.runtime.customized;
 
+import javax.transaction.Synchronization;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
 import javax.transaction.TransactionManager;
 import javax.transaction.UserTransaction;
 
-import org.hibernate.engine.transaction.jta.platform.internal.AbstractJtaPlatform;
+import org.hibernate.engine.transaction.jta.platform.internal.JtaSynchronizationStrategy;
+import org.hibernate.engine.transaction.jta.platform.internal.TransactionManagerAccess;
+import org.hibernate.engine.transaction.jta.platform.internal.TransactionManagerBasedSynchronizationStrategy;
+import org.hibernate.engine.transaction.jta.platform.spi.JtaPlatform;
 
-public final class QuarkusJtaPlatform extends AbstractJtaPlatform {
+public final class QuarkusJtaPlatform implements JtaPlatform, TransactionManagerAccess {
 
     public static final QuarkusJtaPlatform INSTANCE = new QuarkusJtaPlatform();
 
+    private final JtaSynchronizationStrategy tmSynchronizationStrategy = new TransactionManagerBasedSynchronizationStrategy(
+            this);
+    private volatile TransactionManager transactionManager;
+    private volatile UserTransaction userTransaction;
+
     private QuarkusJtaPlatform() {
         //nothing
     }
 
     @Override
-    protected TransactionManager locateTransactionManager() {
-        return com.arjuna.ats.jta.TransactionManager.transactionManager();
+    public TransactionManager retrieveTransactionManager() {
+        TransactionManager transactionManager = this.transactionManager;
+        if (transactionManager == null) {
```

Comment:
Looks like some double checked locking except I don't see the `synchronized` part.

Is it safe?

Reply:
it's safe as the method we invoke is `synchronized` and returns a singleton.